### ISSUE TITLE
Multiple data sources does not work correctly in gsp

### ISSUE THREAD

**Issue by antivista** [grails/grails-core#11645]
### Steps to Reproduce

I have the need to work with two different data sources. So following as described in the doc ([http://docs.grails.org/3.3.5/guide/single.html#dataSource](http://docs.grails.org/3.3.5/guide/single.html#dataSource)), I worked on grails-app/conf/application.yml to configure my second data source. This is how look like my the data source section in my grails-app/conf/application.yml:
```
environments:
    development:
        dataSource:
            dbCreate: update
            username: user1
            password: pass1
            url: jdbc:mysql://localhost:3306/theFirst?zeroDateTimeBehavior=round
        dataSources:
            theSecond:
                #pooled: true
                #transactional: true
                dbCreate: none
                username: user2
                password: pass2
                url: jdbc:mysql://localhost:3306/theSecond?zeroDateTimeBehavior=round
```

I have a domain class OfflineMessage mapped on my second data source in this way:
```
static mapping = {
        datasource 'theSecond'
}
```

A service:
```
@Service(OfflineMessage)
@Transactional(readOnly = true, connection = "theSecond")
interface OfflineMessageService {

    static dataSource = 'theSecond'

    OfflineMessage get(Serializable id)

    List<OfflineMessage> list(Map args)

    Long count()

    void delete(Serializable id)

    OfflineMessage save(OfflineMessage offlineMessage)

}
```

And a OfflineMessageController so defined:
```
@Transactional(readOnly = true)
class OfflineMessageController {

    OfflineMessageService offlineMessageService

    def index(Integer max) {
        params.max = Math.min(max ?: 30, 100)
        respond offlineMessageService.list(params), model:[offlineMessageCount: offlineMessageService.count()]
    }
}
```

I have defined a gsp /offlineMessage/index.gsp where I try to call:
`${OfflineMessage.count}`
but going on this gsp, I have this error:
[1-stacktrace] `Could not obtain current Hibernate Session; nested exception is org.hibernate.HibernateException: No Session found for current thread`
Nevertheless, if I try to call a method of another domain class, which is mapped on theFirst data source, for example:
`${User.count}`
it works correctly. Even calling ${offlineMessageCount} works perfectly.

If I try to specify the data source on OfflineMessageController this way:
`@Transactional(readOnly = true, connection = "theSecond")`
I have this other error:
[2-stacktrace] `DataSource not found for name [theSecond] in configuration. Please check your multiple data sources configuration and try again.`

### Stacktrace
[1-stacktrace]
```
Servlet.service() for servlet [grailsDispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.grails.gsp.GroovyPagesException: Error evaluating expression [OfflineMessage.count] on line [32]: Could not obtain current Hibernate Session; nested exception is org.hibernate.HibernateException: No Session found for current thread] with root cause

org.hibernate.HibernateException: No Session found for current thread
	at org.grails.orm.hibernate.GrailsSessionContext.currentSession(GrailsSessionContext.java:117)
	at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:688)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.getSession(GrailsHibernateTemplate.java:335)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:286)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:243)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:117)
	at org.grails.orm.hibernate.AbstractHibernateGormStaticApi.count(AbstractHibernateGormStaticApi.groovy:152)
	at org.grails.datastore.gorm.GormStaticApi.getCount(GormStaticApi.groovy:548)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper.getCount(GormEntity.groovy:638)
	at Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp$_run_closure2$_closure19.doCall(Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp:53)
	at org.grails.gsp.GroovyPage.evaluate(GroovyPage.java:253)
	at Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp$_run_closure2.doCall(Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp:53)
	at org.grails.taglib.TagBodyClosure.executeClosure(TagBodyClosure.java:200)
	at org.grails.taglib.TagBodyClosure.captureClosureOutput(TagBodyClosure.java:102)
	at org.grails.taglib.TagBodyClosure.call(TagBodyClosure.java:213)
	at org.grails.plugins.web.taglib.SitemeshTagLib.captureTagContent(SitemeshTagLib.groovy:48)
	at org.grails.plugins.web.taglib.SitemeshTagLib$_closure3.doCall(SitemeshTagLib.groovy:156)
	at org.grails.gsp.GroovyPage.invokeTagLibClosure(GroovyPage.java:446)
	at org.grails.gsp.GroovyPage.invokeTag(GroovyPage.java:364)
	at Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp.run(Users_antivista_Documents_web_console_grails_app_views_offlineMessage_index_gsp:60)
	at org.grails.gsp.GroovyPageWritable.doWriteTo(GroovyPageWritable.java:162)
	at org.grails.gsp.GroovyPageWritable.writeTo(GroovyPageWritable.java:82)
	at org.grails.web.servlet.view.GroovyPageView.renderTemplate(GroovyPageView.java:76)
	at org.grails.web.servlet.view.AbstractGrailsView.renderWithinGrailsWebRequest(AbstractGrailsView.java:71)
	at org.grails.web.servlet.view.AbstractGrailsView.renderMergedOutputModel(AbstractGrailsView.java:55)
	at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:303)
	at org.grails.web.sitemesh.GrailsLayoutView.renderInnerView(GrailsLayoutView.java:150)
	at org.grails.web.sitemesh.GrailsLayoutView.obtainContent(GrailsLayoutView.java:128)
	at org.grails.web.sitemesh.GrailsLayoutView.renderTemplate(GrailsLayoutView.java:63)
	at org.grails.web.servlet.view.AbstractGrailsView.renderWithinGrailsWebRequest(AbstractGrailsView.java:71)
	at org.grails.web.servlet.view.AbstractGrailsView.renderMergedOutputModel(AbstractGrailsView.java:55)
	at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:303)
	at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1286)
	at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1041)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:984)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at org.springframework.security.web.session.ConcurrentSessionFilter.doFilter(ConcurrentSessionFilter.java:155)
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at grails.plugin.springsecurity.web.UpdateRequestContextHolderExceptionTranslationFilter.doFilter(UpdateRequestContextHolderExceptionTranslationFilter.groovy:64)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:54)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at com.antivista.ApiBlockingFilter.doFilter(ApiBlockingFilter.groovy:39)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:64)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at com.antivista.security.NetworkInterfaceFilter.doFilterInternal(NetworkInterfaceFilter.groovy:88)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

[2-stacktrace]
```
ConfigurationException occurred when processing request: [GET] /offlineMessage/index
DataSource not found for name [theSecond] in configuration. Please check your multiple data sources configuration and try again.. Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:861)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at org.springframework.security.web.session.ConcurrentSessionFilter.doFilter(ConcurrentSessionFilter.java:155)
	at org.springframework.boot.web.filter.ApplicationContextHeaderFilter.doFilterInternal(ApplicationContextHeaderFilter.java:55)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at grails.plugin.springsecurity.web.UpdateRequestContextHolderExceptionTranslationFilter.doFilter(UpdateRequestContextHolderExceptionTranslationFilter.groovy:64)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.filter.GrailsAnonymousAuthenticationFilter.doFilter(GrailsAnonymousAuthenticationFilter.groovy:54)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at com.antivista.security.ApiBlockingFilter.doFilter(ApiBlockingFilter.groovy:39)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter.doFilter(MutableLogoutFilter.groovy:64)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at grails.plugin.springsecurity.web.SecurityRequestHolderFilter.doFilter(SecurityRequestHolderFilter.groovy:58)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at com.antivista.security.NetworkInterfaceFilter.doFilterInternal(NetworkInterfaceFilter.groovy:88)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.grails.datastore.mapping.core.exceptions.ConfigurationException: DataSource not found for name [theSecond] in configuration. Please check your multiple data sources configuration and try again.
	at org.grails.orm.hibernate.HibernateDatastore.getDatastoreForConnection(HibernateDatastore.java:351)
	at org.grails.orm.hibernate.HibernateDatastore$2.getDatastoreForConnection(HibernateDatastore.java:148)
	... 42 common frames omitted
```

### Environment Information

- Operating System: macOS Catalina 10.15.4
- Grails Version: 3.3.5
- Hibernate Version: 5.1.5
- GORM Version: 6.1.9.RELEASE
- JDK Version: 1.8.0_231

### Description

Do have I missed something on the configuration of the second data source (theSecond)? Even if the thing which surprises me more is: how is it possible I can work with my second data source out of the the gsp? 

Additional note: if I try to work with sessionFactory_theSecond (injecting it), I can not get the current session, opening a new session, I can work on it with any problem.

I hope someone can explain me what I am doing wrong, thanks in advance.

**Comment by jjelliott** [comment]
Grails only does OSIV for the primary datasource. 

If you want to enable OSIV for the secondary datasource, add this bean:
```
openSessionInViewInterceptor_theSecond(GrailsOpenSessionInViewInterceptor) {
    sessionFactory = ref("sessionFactory_theSecond")
}
```

_*However*_, you're mixing two approaches. You are using GORM directly as well as using a data service. Instead of calling `OfflineMessage.count()`, you should be using the `offlineMessageCount` variable you're passing up from the controller.

**Comment by antivista** [comment]
I did not know that Grails does OSIV only for the primary datasource.
Thanks for this valuable info and for the help to enable OSIV for a secondary datasource, I have tried to enable it as you described (in my grails-app/conf/spring/resources.groovy in the beans closure) and it works perfectly.
I am experimenting with both approaches to be able to use one or the other depending on the situation, it was just a try, for this reason in `OfflineMessageController` there was that `offlineMessageCount`.

Really thank you!
